File: 202501/2948.py
class Solution:
    def lexicographicallySmallestArray(self, nums: List[int], limit: int) -> List[int]:
        n = len(nums)
        group = [-1] * n
        a = list((v, k) for k, v in enumerate(nums))
        a.sort()
        groups = []
        cg = []
        value = []
        for i in range(n):
            if len(cg) > 0 and a[i - 1][0] + limit < a[i][0]:
                for j in cg:
                    group[j] = len(groups)
                groups.append(sorted(value))
                cg = []
                value = []
            cg.append(a[i][1])
            value.append(a[i][0])
        for j in cg:
            group[j] = len(groups)
        groups.append(sorted(value))
        idx = [0] * len(groups)

        ans = [0] * n
        for i in range(n):
            ans[i] = groups[group[i]][idx[group[i]]]
            idx[group[i]] += 1

        return ans

File: 202608/2948.cpp
class Solution {
public:
  vector<int> lexicographicallySmallestArray(vector<int> &nums, int limit) {
    vector<int> sorted(nums);
    sort(begin(sorted), end(sorted));
    int n = nums.size();
    unordered_map<int, vector<int>> groups;
    unordered_map<int, int> item_2_groups;
    int group = -1;
    for (int i = 0; i < n; ++i) {
      if (i == 0 || sorted[i] > sorted[i - 1] + limit) {
        ++group;
      }
      item_2_groups[sorted[i]] = group;
      groups[group].push_back(sorted[i]);
    }
    vector<int> ans(n);
    unordered_map<int, int> group_idx;
    for (int i = 0; i < n; ++i) {
      group = item_2_groups[nums[i]];
      ans[i] = groups[group][group_idx[group]++];
    }
    return ans;
  }
};
